Sources/Sentry/SentryMetricProfiler.mm

```diff
@@ -0,0 +1,156 @@
+        _thermalStateChanges = [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+        _memoryPressureStateChanges =
+            [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+        _notificationCenter = notificationCenterWrapper;
+        _profileStartTime = profileStartTime;
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    [self stop];
+}
+
+#pragma mark - Public
+
+- (void)start
+{
+    [self registerTimeSeriesHandler];
+    [self registerMemoryPressureWarningHandler];
+}
+
+- (void)stop
+{
+    [_timer invalidate];
+    dispatch_source_cancel(_memoryWarningSource);
+    [_notificationCenter removeObserver:self name:NSProcessInfoThermalStateDidChangeNotification];
+}
+
+- (NSData *)serialize
+{
+    // TODO: implement
+    return [[NSData alloc] init];
+}
+
+#pragma mark - Private
+
+- (void)registerTimeSeriesHandler
+{
+    _timer = [NSTimer scheduledTimerWithTimeInterval:kSentryMetricProfilerInterval
+                                             repeats:YES
+                                               block:^(NSTimer *_Nonnull timer) {
+                                                   [self recordCPUPercentage];
+                                                   [self recordMemoryFootprint];
+                                               }];
+}
+
+/**
+ * This is a more fine-grained API, providing normal/warn/critical levels of memory usage, versus
+ * using `UIApplicationDidReceiveMemoryWarningNotification` which does not provide any additional
+ * information ("This notification does not contain a userInfo dictionary." from
+ * https://developer.apple.com/documentation/uikit/uiapplication/1622920-didreceivememorywarningnotificat).
+ */
+- (void)registerMemoryPressureWarningHandler
+{
+    const auto queueAttributes
+        = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_UTILITY, 0);
+    _memoryWarningQueue = dispatch_queue_create("io.sentry.queue.memory-warnings", queueAttributes);
+    _memoryWarningSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_MEMORYPRESSURE, 0,
+        DISPATCH_MEMORYPRESSURE_NORMAL | DISPATCH_MEMORYPRESSURE_WARN
+            | DISPATCH_MEMORYPRESSURE_CRITICAL,
+        _memoryWarningQueue);
+    dispatch_source_set_event_handler(_memoryWarningSource, ^{
+        [self recordMemoryPressureState:dispatch_source_get_data(self->_memoryWarningSource)];
+    });
+    dispatch_resume(_memoryWarningSource);
+}
+
+- (void)registerStateChangeNotifications
+{
+    // According to Apple docs: "To receive NSProcessInfoThermalStateDidChangeNotification, you must
+    // access the thermalState prior to registering for the notification." (from
+    // https://developer.apple.com/documentation/foundation/nsprocessinfothermalstatedidchangenotification/)
+    [self recordThermalState];
+
+    // According to Apple docs: "This notification is posted on the global dispatch queue."
+    [_notificationCenter addObserver:self
```

Comment:
`m`: You can replace this with our wrapper `SentryNSNotificationCenterWrapper`, pass in `TestNSNotificationCenterWrapper` as a dependency, and then you can test this properly.